app/src/main/java/com/techcourse/DispatcherServlet.java

```diff
@@ -1,51 +1,139 @@
+            log.info("initialize HandlerMapping :: %s".formatted(handlerMapping.getClass().getName()));
+            handlerMapping.initialize(); // 초기화 (매핑되는 객체들 생성)
+        }
+    }
+
+    /**
+     * HTTP 요청을 처리
+     * 1. 요청에 매핑된 핸들러가 있는지 조회
+     * 2. 핸들러를 실행할 어댑터 조회
+     * 3. 핸들러 실행
+     * 4. 뷰 렌더링
+     */
     @Override
     protected void service(final HttpServletRequest request, final HttpServletResponse response) throws ServletException {
         final String requestURI = request.getRequestURI();
         log.debug("Method : {}, Request URI : {}", request.getMethod(), requestURI);
 
         try {
-            final var controller = manualHandlerMapping.getHandler(requestURI);
-            final var viewName = controller.execute(request, response);
-            move(viewName, request, response);
+            Object handler = getHandler(request); // 1. 요청에 매핑된 핸들러 조회
+            HandlerAdaptor handlerAdaptor = getAdaptor(handler); // 2. 핸들러를 실행할 어댑터 조회
+
+            ModelAndView modelAndView = handlerAdaptor.handle(request, response, handler); // 3. 핸들러 실행 후 ModelAndView 반환
+            render(modelAndView, request, response); // 4. View 렌더링 수행
         } catch (Throwable e) {
             log.error("Exception : {}", e.getMessage(), e);
             throw new ServletException(e.getMessage());
         }
     }
 
-    private void move(final String viewName, final HttpServletRequest request, final HttpServletResponse response) throws Exception {
-        if (viewName.startsWith(JspView.REDIRECT_PREFIX)) {
-            response.sendRedirect(viewName.substring(JspView.REDIRECT_PREFIX.length()));
-            return;
+    /**
+     * 핸들러에 맞는 HandlerAdaptor를 찾는 메서드.
+     * 각 어댑터의 supports() 메서드를 통해 처리 가능 여부 판단
+     * @param handler 실행할 핸들러 객체
+     * @return 해당 핸들러를 처리할 수 있는 어댑터
+     * @throws IllegalArgumentException 지원하는 어댑터가 없으면 예외 던짐
+     */
+    private HandlerAdaptor getAdaptor(Object handler) {
+        for (HandlerAdaptor handlerAdaptor : handlerAdaptors) {
+            if(handlerAdaptor.supports(handler)) {
+                return handlerAdaptor;
+            }
+        }
+        throw new IllegalArgumentException("매핑 가능한 HandlerAdaptor가 존재하지 않습니다. (Handler type: %s)"
+                .formatted(handler.getClass().getName()));
+    }
+
+    /**
+     * 요청에 맞는 핸들러를 찾는 메서드.
+     * 등록된 HandlerMapping 리스트에서 순차적으로 요청 URI에 매핑되는 핸들러 조회
+     * @param request 클라이언트 요청 객체
+     * @return 요청을 처리할 핸들러 객체
+     * @throws IllegalArgumentException 매핑 가능한 핸들러가 없으면 예외 던짐
+     */
+    private Object getHandler(HttpServletRequest request) {
+        for (HandlerMapping handlerMapping : handlerMappings) {
+            Object handler = handlerMapping.getHandler(request);
+            if(handler != null) {
+                return handler;
+            }
         }
+        throw new IllegalArgumentException("매핑 가능한 Handler가 존재하지 않습니다. (Request URI : %s"
+                .formatted(request.getRequestURI()));
```

Comment:
아래처럼 바꿔볼 수도 있을 것 같아요!

```suggestion
handlerMappings.stream()
            .map(handlerMapping -> handlerMapping.getHandler(request))
            .filter(Objects::nonNull)
            .findFirst()
            .orElseThrow(() -> throw new IllegalArgumentException("매핑 가능한 Handler가 존재하지 않습니다. (Request URI : %s".formatted(request.getRequestURI()));
```